planner/core/task.go

```diff
@@ -955,14 +946,61 @@ func (p *PhysicalTopN) canPushToIndexPlan(indexPlan PhysicalPlan, byItemCols []*
 	return true
 }
 
+// canExpressionConvertedToPB checks whether each of the the expression in TopN can be converted to pb.
+func (p *PhysicalTopN) canExpressionConvertedToPB(storeTp kv.StoreType) bool {
+	exprs := make([]expression.Expression, 0, len(p.ByItems))
+	for _, item := range p.ByItems {
+		exprs = append(exprs, item.Expr)
+	}
+	return expression.CanExprsPushDown(p.ctx.GetSessionVars().StmtCtx, exprs, p.ctx.GetClient(), storeTp)
+}
+
+// containVirtualColumn checks whether TopN.ByItems contains virtual generated columns.
+func (p *PhysicalTopN) containVirtualColumn(copTask *copTask) bool {
+	tCols := copTask.tablePlan.Schema().Columns
+	for _, by := range p.ByItems {
+		cols := expression.ExtractColumns(by.Expr)
+		for _, col := range cols {
+			if col.VirtualExpr != nil {
+				for _, tCol := range tCols {
+					// A column with ID > 0 indicates that the column can be resolved by data source.
+					if tCol.ID > 0 && tCol.ID == col.ID && tCol.VirtualExpr != nil {
+						return true
+					}
+				}
+			}
+		}
+	}
+	return false
+}
+
+// canPushDownToTiKV checks whether this topN can be pushed down to TiKV.
+func (p *PhysicalTopN) canPushDownToTiKV(copTask *copTask) bool {
+	if !p.canExpressionConvertedToPB(kv.TiKV) {
+		return false
+	}
+	if len(copTask.rootTaskConds) != 0 {
+		return false
+	}
+	if p.containVirtualColumn(copTask) {
+		return false
+	}
+	return true
+}
+
+// canPushDownToTiFlash checks whether this topN can be pushed down to TiFlash.
+func (p *PhysicalTopN) canPushDownToTiFlash() bool {
+	return p.canExpressionConvertedToPB(kv.TiFlash)
+}
+
 func (p *PhysicalTopN) attach2Task(tasks ...task) task {
```

Comment:
Just a question(may not related to this pr): do we have test cases to make sure other operatores(like Selection, Projection etc) using virtual column will not pushdown to tiflash/tikv.

Reply:
make sense, can we just add the virtual column check in `columnToPBExpr <- ExpressionToPBExpr` which in unified judgement of all Operator/Expression about whether they can be pushd down to kv to tiflash

Reply:
which also will reduce/simplify your code here I think